### PR TITLE
feat: add product slug to links to the PDP

### LINKED ISSUE #11489: Product slug for a better product title in the URL
While adding a pretty product title has been easy to implement by merchants, we like to make this a standard feature. 

Product slug is often something you can configure in the backend, but in case of product titles there's no point in manually creating the slug for each product, which is why we generate it: 
- trim the titles
- lower case the title
- avoid encoded characters; replace space, comma and plus characters with dashes
- avoid multiple dashes

We also like to ensure the merchant can customise the conversion. We'll not offer configuration in the first take, but that could be something to consider going forward. This will be part of the adapter converters, so in case a product backend service provides a slug, it can work transparent in Spartacus.

### PR DESCRIPTION
**Pull request by tobi-or-not-tobi**
closes #11489

### CODE REVIEW

File: projects/core/src/occ/adapters/product/converters/product-name-normalizer.ts [hunk 1]
```diff
@@ -1,25 +1,38 @@
...
 
 @Injectable({ providedIn: 'root' })
 export class ProductNameNormalizer implements Converter<Occ.Product, Product> {
   constructor(protected config: OccConfig) {}
 
   convert(source: Occ.Product, target?: Product): Product {
-    if (target === undefined) {
-      target = { ...(source as any) };
-    }
+    target = target ?? { ...((source as any) as Partial<Product>) };
```
Comment (znikola) [inline projects/core/src/occ/adapters/product/converters/product-name-normalizer.ts]: According to our TS expert, @marlass, we should avoid using `any`, and use `unknown` instead.

File: projects/core/src/occ/adapters/product/converters/product-name-normalizer.ts [hunk 2]
```diff
@@ -1,25 +1,38 @@
...
+
+  /**
+   * Provides a title slug for the pretty URL.
+   *
+   * The name is created in lowercase, trimmed and all special characters that are candidates for url encoded
+   * are replaced by the slug char (dash by default).
+   */
+  protected normalizeSlug(name: string): string {
+    // https://developers.google.com/maps/documentation/urls/url-encoding
+    const reservedChars = ` !*'();:@&=+$,/?%#[]`;
```
Comment (znikola) [inline projects/core/src/occ/adapters/product/converters/product-name-normalizer.ts]: Should we move this const outside of the method?
Not sure if the compiler will optimize this, but we could potentially avoid re-initializing this const every time the `normalizeSlug` is called.
One other benefit could be the extensibility - customers would be able to override it.

File: projects/core/src/occ/adapters/product/converters/product-name-normalizer.ts [hunk 3]
```diff
@@ -1,25 +1,38 @@
...
+    const re = new RegExp(`[${reservedChars.split('').join('\\')}]`, 'g');
```
Comment (znikola) [inline projects/core/src/occ/adapters/product/converters/product-name-normalizer.ts]: Similar to the above - we could instantiate the `new RegExp` only once in this class, and avoid having to instantiate it every time the method is called.

**Review by znikola — CHANGES_REQUESTED** [review]

File: projects/core/src/occ/adapters/product/converters/product-name-normalizer.ts (hunk 2 above)
Comment (tobi-or-not-tobi) [reply projects/core/src/occ/adapters/product/converters/product-name-normalizer.ts]: Agreed. i had that initially, then strugled with regex and moved it out. I'll move it in again.

**Review by znikola — APPROVED** [review]

**Pull request merged**